### PR TITLE
Optimizer-in-backward Fusion Implementation

### PR DESCRIPTION
**Pull request by tinuademargaret**
This PR implements optimizer-in-backward fusion (suggested in this [issue](https://github.com/volcengine/verl/issues/144)), which integrates optimizer updates directly into the backward pass to decrease peak memory usage by eliminating the need to store gradients. It uses pytorch's  `_apply_optimizer_in_backward `to fuse optimizer steps with backpropagation.

This feature only works with the sft trainer. 

Optimizer-in-backward fusion is incompatible with gradient accumulation and is controlled by a `bwd_hook` flag in the optimizer config.

There is added validation to ensure that training configurations that rely on gradient accumulation does not enable the optimizer fusion option.

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
thanks! Just to double check, this supports gradient accumulation?

**Comment by tinuademargaret** [comment]
> thanks! Just to double check, this supports gradient accumulation?

 No, It uses pytorch's `apply_optimizer_in_backward`  which does not support gradient accumulation.

**Review by eric-haibin-lin — CHANGES_REQUESTED** [review]
Thanks for the contribution! For RL i think it's very likely that we need gradient accumulation to train with large batch sizes. But this feature would still be useful for SFT. Do you think it make sense if you only keep the changes for SFT trainer, while keeping the RL trainer simple for now? Thx
